buffer/buffer.go

```diff
@@ -0,0 +1,111 @@
+// Package buffer implements a variable-sized bytes pool.
+package buffer
+
+import (
+	"sync"
+)
+
+const (
+	numBuckets = 256
+	pageSize   = 1024
+)
+
+// Buffers contains buckets for sharding.
+type Buffers struct {
+	pageSize int
+	buckets  [numBuckets]bucket
+}
+
+type bucket struct {
+	lock  sync.RWMutex
+	pools map[int]*Pool
+}
+
+// Pool represents a fixed size bytes pool.
+type Pool struct {
+	size int
+	pool *sync.Pool
+}
+
+// GetBuffer returns a bytes from the pool with the given size.
+func (p *Pool) GetBuffer(size int) []byte {
+	return p.pool.Get().([]byte)[:size]
+}
+
+// PutBuffer frees the bytes to the pool.
+func (p *Pool) PutBuffer(buf []byte) {
+	buf = buf[:cap(buf)]
+	if cap(buf) >= p.size {
+		p.pool.Put(buf)
+	}
+}
+
+// NewBuffers creates a new Buffers with the given page size.
+func NewBuffers(pageSize int) *Buffers {
+	b := new(Buffers)
+	for i := range b.buckets {
+		b.buckets[i].pools = make(map[int]*Pool)
+	}
+	if pageSize < 1 {
+		b.pageSize = 1
+	} else {
+		b.pageSize = pageSize
+	}
+	return b
+}
+
+// AssignPool assigns a fixed size bytes pool with the given size.
+func (b *Buffers) AssignPool(size int) (p *Pool) {
```

Comment:
Can we make it unexported?

Reply:
There is a time cost to assign a bytes pool. 
When we need to put the buffer immediately after using it.
```
p := buffer.AssignPool(size)
buf = p.GetBuffer(size)
p.PutBuffer(buf)
```
faster than
```
buf = buffer.GetBuffer(size)
buffer.PutBuffer(buf)
```